app/src/main/java/co/stonephone/stonecamera/StoneCameraViewModel.kt

```diff
@@ -351,28 +363,47 @@ class StoneCameraViewModel(
 
                 val cameraSelector = createCameraSelectorForId(_selectedCameraId!!)
 
+                val manager = MyApplication.getAppContext()
+                    .getSystemService(Context.CAMERA_SERVICE) as CameraManager
+                val chars = manager.getCameraCharacteristics(_selectedCameraId!!)
+                val level = chars.get(CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL)
+                val numConcurrentUseCases = when (level) {
+                    CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_LEGACY -> 1
+                    CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_LIMITED -> 2
+                    CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_FULL -> 3
+                    CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_3 -> 3
+                    else -> 1
+                }
+
+
                 previewViewTouchHandlers.clear()
                 _cameraProvider!!.unbindAll()
 
-                // TODO move this into a plugin-level solution
-                // also: on more powerful devices that can support 3 use-cases, we should bind them all from day 1 for fast switching
-                if(selectedMode == "Photo") {
-                    camera = _cameraProvider!!.bindToLifecycle(
-                        lifecycleOwner!!,
-                        cameraSelector,
-                        preview,
-                        imageCapture,
-                        imageAnalysis,
-                    )
-                } else {
-                    camera = _cameraProvider!!.bindToLifecycle(
-                        lifecycleOwner!!,
-                        cameraSelector,
-                        preview,
-                        videoCapture,
-                    )
+                val selectedModePlugin = plugins.find { it.modeLabel == selectedMode }
+                val requiredUseCases: List<PluginUseCase> =
+                    selectedModePlugin?.modeUseCases ?: emptyList()
+
+                // All use-cases (including not needed by selected plugin), but retaining the order from the selected plugin
```

Comment:
v.minor suggested change to the comment to improve the grammar: "All use-cases (including those not needed by selected plugin), retaining the order of the selected plugin"